### PR TITLE
build: use build interface for the case of add_subdirectory

### PR DESCRIPTION
**Pull request by gocarlos**
this adds build interface to the target include directories, which is important when using add_subdirectory in cmake

### CODE REVIEW

**Comment by gocarlos** [comment]
related to https://github.com/ARMmbed/mbedtls/issues/3108

File: library/CMakeLists.txt [hunk 1]
```diff
@@ -189,21 +199,27 @@ if(USE_SHARED_MBEDTLS_LIBRARY)
...
+        PUBLIC
+            $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/../include>
+            $<INSTALL_INTERFACE:include>)
 
     add_library(mbedtls SHARED ${src_tls})
     set_target_properties(mbedtls PROPERTIES VERSION 2.21.0 SOVERSION 13)
     target_link_libraries(mbedtls ${libs} mbedx509)
-    target_include_directories(mbedtls
-        PUBLIC ${MBEDTLS_DIR}/include/)
-
```
Comment (gilles-peskine-arm) [inline library/CMakeLists.txt:206]: Please remove trailing whitespace and run `tests/scripts/check-files.py` before pushing.

File: library/CMakeLists.txt [hunk 2]
```diff
@@ -189,21 +199,27 @@ if(USE_SHARED_MBEDTLS_LIBRARY)
     add_library(mbedcrypto SHARED ${src_crypto})
     set_target_properties(mbedcrypto PROPERTIES VERSION 2.21.0 SOVERSION 4)
     target_link_libraries(mbedcrypto ${libs})
-    target_include_directories(mbedcrypto
-        PUBLIC ${MBEDTLS_DIR}/include/)
+    target_include_directories(mbedcrypto 
+        PUBLIC
+            $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/../include>
```
Comment (gilles-peskine-arm) [inline library/CMakeLists.txt:204]: What's `$<…>`? I can't find it in https://cmake.org/cmake/help/v3.0/manual/cmake-language.7.html#syntax

Comment (gocarlos) [reply library/CMakeLists.txt:204]: https://cmake.org/cmake/help/v3.3/manual/cmake-generator-expressions.7.html

File: library/CMakeLists.txt (hunk 1 above)
Comment (gocarlos) [reply library/CMakeLists.txt:206]: in the future to automatically format cmake one could introduce: https://pypi.org/project/cmake-format/
just `pip install cmake-format`

**Comment by gocarlos** [comment]
> What doesn't currently work, and works after this change?

from https://cmake.org/cmake/help/v3.0/command/target_include_directories.html:

```
Include directories usage requirements commonly differ between the build-tree and the install-tree. The BUILD_INTERFACE and INSTALL_INTERFACE generator expressions can be used to describe separate usage requirements based on the usage location. Relative paths are allowed within the INSTALL_INTERFACE expression and are interpreted relative to the installation prefix. For example:

target_include_directories(mylib PUBLIC
  $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include/mylib>
  $<INSTALL_INTERFACE:include/mylib>  # <prefix>/include/mylib
)
```

File: library/CMakeLists.txt (hunk 1 above)
Comment (gilles-peskine-arm) [reply library/CMakeLists.txt:206]: Good idea. We've been considering `clang-format` for C code, good to know that there's something similar for cmake.

File: library/CMakeLists.txt (hunk 2 above)
Comment (gilles-peskine-arm) [reply library/CMakeLists.txt:204]: Thanks!

**Comment by gocarlos** [comment]
> This looks sensible to me, but I know very little about cmake.
> 
> Could you please clarify why you're making this change in the commit message? What doesn't currently work, and works after this change?
> 
> Would it be possible to add a non-regression test, so that our continuous integration will tell us if we break this new feature? We don't yet test `make install`, but we should, and now would be a good time to test it. The right place for this test would be in `tests/scripts/all.sh`, either in `component_test_cmake_out_of_source` or in `component_test_cmake_as_subdirectory` or in a new component.

I think its pretty hard to add tests to this, as the issues I had were together with cross compilation and yocto... to add an yocto test would be pretty big effort...

**Comment by bensze01** [comment]
> > This looks sensible to me, but I know very little about cmake.
> > Could you please clarify why you're making this change in the commit message? What doesn't currently work, and works after this change?
> > Would it be possible to add a non-regression test, so that our continuous integration will tell us if we break this new feature? We don't yet test `make install`, but we should, and now would be a good time to test it. The right place for this test would be in `tests/scripts/all.sh`, either in `component_test_cmake_out_of_source` or in `component_test_cmake_as_subdirectory` or in a new component.
> 
> I think its pretty hard to add tests to this, as the issues I had were together with cross compilation and yocto... to add an yocto test would be pretty big effort...

While your patch seems fine to me, and from my limited testing produces correct results - as in, the same install results as the code pre-patch - we would very much like to have a test for this bug before we merge your PR.

Could you help us with producing a minimal example that produces different results with and without your patch applied?
Or provide a bit more detail about your setup / steps to reproduce the bug, so that we can develop one ourselves?

**Comment by gocarlos** [comment]
> > > This looks sensible to me, but I know very little about cmake.
> > > Could you please clarify why you're making this change in the commit message? What doesn't currently work, and works after this change?
> > > Would it be possible to add a non-regression test, so that our continuous integration will tell us if we break this new feature? We don't yet test `make install`, but we should, and now would be a good time to test it. The right place for this test would be in `tests/scripts/all.sh`, either in `component_test_cmake_out_of_source` or in `component_test_cmake_as_subdirectory` or in a new component.
> > 
> > 
> > I think its pretty hard to add tests to this, as the issues I had were together with cross compilation and yocto... to add an yocto test would be pretty big effort...
> 
> While your patch seems fine to me, and from my limited testing produces correct results - as in, the same install results as the code pre-patch - we would very much like to have a test for this bug before we merge your PR.
> 
> Could you help us with producing a minimal example that produces different results with and without your patch applied?
> Or provide a bit more detail about your setup / steps to reproduce the bug, so that we can develop one ourselves?

https://stackoverflow.com/a/58729533

**Comment by ronald-cron-arm** [comment]
Thanks for the pointer to stackoverflow. I have added an install(EXPORT) command for the libraries (see https://github.com/ronald-cron-arm/mbedtls/tree/install_export) and I can see an error in the generate step of cmake:

CMake Error in library/CMakeLists.txt:
  Target "mbedcrypto" INTERFACE_INCLUDE_DIRECTORIES property contains path:
    "/home/roncro01/work/mbedtls/include/"
  which is prefixed in the source directory.

And this error disappears when applying your commit "build: use build interface for the case of add_subdirectory".  We may be able to build a test on that. Would this catch well what your commit intends to fix? Otherwise I am wondering if it would be useful to add permanently this install(EXPORT) command to our cmake build system?

**Comment by gilles-peskine-arm** [comment]
Please note that there are merge conflicts.

**Comment by ronald-cron-arm** [comment]
This PR has been stalled for a few months thus we are closing it. Please re-open it if you want to resume your work on it.

**Pull request closed without merging**